config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java

```diff
@@ -0,0 +1,237 @@
+import org.springframework.security.oauth2.server.resource.web.access.BearerTokenAccessDeniedHandler;
+import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+
+/**
+ *
+ * An {@link AbstractHttpConfigurer} for OAuth 2.0 Resource Server Support.
+ *
+ * By default, this wires a {@link BearerTokenAuthenticationFilter}, which can be used to parse the request
+ * for bearer tokens and make an authentication attempt.
+ *
+ * <p>
+ * The following configuration options are available:
+ *
+ * <ul>
+ * <li>{@link #jwt()} - enables Jwt-encoded bearer token support</li>
+ * </ul>
+ *
+ * <p>
+ * When using {@link #jwt()}, a Jwk Set Uri must be supplied via {@link JwtConfigurer#jwkSetUri}
+ *
+ * <h2>Security Filters</h2>
+ *
+ * The following {@code Filter}s are populated when {@link #jwt()} is configured:
+ *
+ * <ul>
+ * <li>{@link BearerTokenAuthenticationFilter}</li>
+ * </ul>
+ *
+ * <h2>Shared Objects Created</h2>
+ *
+ * The following shared objects are populated:
+ *
+ * <ul>
+ * <li>{@link SessionCreationPolicy} (optional)</li>
+ * </ul>
+ *
+ * <h2>Shared Objects Used</h2>
+ *
+ * The following shared objects are used:
+ *
+ * <ul>
+ * <li>{@link AuthenticationManager}</li>
+ * </ul>
+ *
+ * If {@link #jwt()} isn't supplied, then the {@link BearerTokenAuthenticationFilter} is still added, but without
+ * any OAuth 2.0 {@link AuthenticationProvider}s. This is useful if needing to switch out Spring Security's Jwt support
+ * for a custom one.
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ * @see BearerTokenAuthenticationFilter
+ * @see JwtAuthenticationProvider
+ * @see NimbusJwtDecoderJwkSupport
+ * @see AbstractHttpConfigurer
+ */
+public final class OAuth2ResourceServerConfigurer<H extends HttpSecurityBuilder<H>> extends
+		AbstractHttpConfigurer<OAuth2ResourceServerConfigurer<H>, H> {
+
+	private BearerTokenResolver resolver = new DefaultBearerTokenResolver();
+	private BearerTokenRequestMatcher matcher = new BearerTokenRequestMatcher();
+
+	private BearerTokenAuthenticationEntryPoint entryPoint = new BearerTokenAuthenticationEntryPoint();
+	private BearerTokenAccessDeniedHandler deniedHandler = new BearerTokenAccessDeniedHandler();
+
+	private JwtConfigurer jwtConfigurer;
+
+	public JwtConfigurer jwt() {
+		if (this.jwtConfigurer == null) {
+			this.jwtConfigurer = new JwtConfigurer();
+		}
+
+		return this.jwtConfigurer;
+	}
+
+	@Override
+	public void setBuilder(H http) {
+		super.setBuilder(http);
+		initSessionCreationPolicy(http);
+	}
+
+	@Override
+	public void init(H http) throws Exception {
+		registerDefaultAccessDeniedHandler(http);
+		registerDefaultEntryPoint(http);
+		registerDefaultCsrfOverride(http);
+	}
+
+	@Override
+	public void configure(H http) throws Exception {
+		BearerTokenResolver resolver = getBearerTokenResolver(http);
+		resolver = postProcess(resolver);
```

Comment:
We want to avoid invoking postProcess on beans a user may have injected since this would cause issues (i.e. their lifecycle methods would be invoked twice).

The current code is a bit fragile in this respect because it is likely that in the future a resolver will be able to be registered by a user explicitly. If this happens, we would likely forget to only postProcess the default value.

We should only invoke postProcess right next to a new statement within a configurer.

Reply:
Okay, great. In hindsight, I had intended to ask you when postProcess is useful. Your explanation makes it clear.